Explicit assignment operators for Observation, default assignment is a bug here

topin89 opened the pull request:
Also move constructor. Default copy assignment is not just a shortcut to copy constructor, sadly

<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket this addresses   | #2412) |
| Primary OS tested on | Ubuntu under WSL1 |
| Robotic platform tested on | N/A |

---

## Description of contribution in a few bullet points

* Fixed bug with copy assignment being just a copy of all class fields, without proper `new` call for `sensor_msgs::msg::PointCloud2 * cloud_`
* Added move assignment and move constructor, because it's they are called inside `std::vector` with less (de)allocations

<!--
* I added this neat new feature
* Also fixed a typo in a parameter name in nav2_costmap_2d
-->

## Description of documentation updates required from your changes
Teeeny changes:
* Changed description for copy assignment
* Added description for move assignment and move constructor
<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

---

## Future work that may be required in bullet points
* I guess similar pulls for `crystal` and `main` braches should be done
* Maybe check if any other default defined functions really shouldn't be
<!--
* I think there might be some optimizations to be made from STL vector
* I see alot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/observation.hpp:
```diff
@@ -89,6 +112,17 @@ class Observation
+   * @brief  Move constructor
+   * @param obs The observation to move
+   */
+  Observation(Observation && obs)
```
This seems like significant overkill. These are `double`s, using std::move on them is not necessary as far as I can tell

topin89 replied:
Yes, looks like nothing here needs `std::move` for performance. Mostly here for uniformity and to show that everything is moving here.

> Added move assignment and move constructor, because it's they are called inside `std::vector` with less (de)allocations

That's why I think move constructor (as well as assignment) is needed, so there is no unnecessary PointCloud copies/allocations/deallocations every time some vector needs to grow.

SteveMacenski replied:
We want things to be created as needed and not have dead code in the system. If its not being used and isn't needed, then it should not be added. Its more code then we (I) need to maintain.

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/observation.hpp:
```diff
@@ -60,9 +60,32 @@ class Observation
+  * @brief  Move assignment operator
+  * @param obs The observation to move
+  */
+  Observation & operator=(Observation && obs){
```
As well here, this is does not seem necessary

topin89 replied:
> Added move assignment and move constructor, because it's they are called inside `std::vector` with less (de)allocations

 Just like up there. I'm not sure what exactly needed, constructor or assignment, so why not both?

SteveMacenski replied:
We want things to be created as needed and not have dead code in the system. If its not being used and isn't needed, then it should not be added. Its more code then we (I) need to maintain.

SteveMacenski reviewed (changes requested):
Overall, I don't see the value in this PR. What are you trying to accomplish that was not possible prior, specifically?

SteveMacenski commented on nav2_costmap_2d/include/nav2_costmap_2d/observation.hpp:
```diff
@@ -60,9 +60,32 @@ class Observation
+  * @brief  Copy assignment operator
+  * @param obs The observation to copy
+  */
+  Observation & operator=(const Observation & obs){
```
This looks fine to me, but I don't think any of the other changes are required.

topin89 replied:
They're indeed not required, but they improve performance a bit.

topin89 commented:
I guess there's an alternative to those verbose constructors/assignments. I didn't try to compile it, but it looks like
`sensor_msgs::msg::PointCloud2 * cloud_;` can be easily replaced with `sensor_msgs::msg::PointCloud2::UniquePtr cloud_;`.

SteveMacenski commented:
Yes, a unique pointer would also work. I'd accept a PR to do that if you like. But please do remove the other unnecessary changes and lets just focus on the copy operator that isn't handling the PC2 correctly.

topin89 commented:
Not sure what exactly should I do, so for now just removed functions unrelated to the bug and fixed linting issues. If I should replace the pointer with `unique_ptr`, I'll change to that. I could also port this to any branches you think it should be ported to, just not sure what to port, how many PRs to do (I mean, I still want to add those move functions). As I understand it, there should be two requests. one for assignment (with or without unique_ptr) and another for move stuff, for every branch.

SteveMacenski reviewed (approved):
Please submit against Galactic and Main branches and I can merge the full set of 3 together

If you wished to switch to a unique pointer as well, that would be fine, but not required. I'm OK with it either way.

topin89 commented:
OK, two followup PRs are done: #2425 #2424

> You will also need to submit a PR https://github.com/ros-planning/navigation.ros.org/blob/master/migration/Galactic.rst against this file adding a new section about this change to the migration guide.

But what to submit there? This is just a bugfix, and there's no difference between `observation.hpp` in `galactic` and `main` branches.

I can think of only
> * Various bug fixes

or
> * Fixed a double free bug 

to "Major Improvements", but that's not exactly major.
